File: golang/proof/did_wba_binding.go
package proof

import (
	"time"

	anp "github.com/agent-network-protocol/anp/golang"
	"github.com/agent-network-protocol/anp/golang/internal/base64util"
	"github.com/agent-network-protocol/anp/golang/internal/diddoc"
)

var DidWbaBindingRequiredFields = []string{
	"agent_did",
	"verification_method",
	"leaf_signature_key_b64u",
	"issued_at",
	"expires_at",
}

// DidWbaBindingVerificationOptions configures business checks around did:wba bindings.
type DidWbaBindingVerificationOptions struct {
	Now                        string
	ExpectedLeafSignatureKey   string
	ExpectedCredentialIdentity string
}

// GenerateDidWbaBinding creates and signs a strict did:wba binding object.
func GenerateDidWbaBinding(agentDID string, verificationMethod string, leafSignatureKeyB64U string, privateKey anp.PrivateKeyMaterial, issuedAt string, expiresAt string, proofCreated string) (map[string]any, error) {
	binding := map[string]any{
		"agent_did":               agentDID,
		"verification_method":     verificationMethod,
		"leaf_signature_key_b64u": leafSignatureKeyB64U,
		"issued_at":               issuedAt,
		"expires_at":              expiresAt,
	}
	if err := validateDidWbaBindingFields(binding); err != nil {
		return nil, err
	}
	if proofCreated == "" {
		proofCreated = issuedAt
	}
	return GenerateObjectProof(binding, privateKey, verificationMethod, agentDID, proofCreated)
}

// VerifyDidWbaBinding verifies a strict did:wba binding object.
func VerifyDidWbaBinding(binding map[string]any, issuerDocument map[string]any, options DidWbaBindingVerificationOptions) error {
	if err := validateDidWbaBindingFields(binding); err != nil {
		return err
	}
	agentDID := stringValue(binding["agent_did"])
	verificationMethod := stringValue(binding["verification_method"])
	if did, err := didFromVerificationMethod(verificationMethod); err != nil || did != agentDID {
		return &Error{Message: "verification_method must belong to agent_did"}
	}
	if !diddoc.IsAssertionMethodAuthorized(issuerDocument, verificationMethod) {
		return &Error{Message: "verification_method is not authorized for assertionMethod"}
	}
	if diddoc.FindVerificationMethod(issuerDocument, verificationMethod) == nil {
		return &Error{Message: "verification_method not found in DID document"}
	}
	if options.ExpectedLeafSignatureKey != "" && options.ExpectedLeafSignatureKey != stringValue(binding["leaf_signature_key_b64u"]) {
		return &Error{Message: "leaf_signature_key_b64u does not match the expected MLS leaf key"}
	}
	if options.ExpectedCredentialIdentity != "" && options.ExpectedCredentialIdentity != agentDID {
		return &Error{Message: "credential.identity does not match agent_did"}
	}
	issuedAt, err := parseRFC3339String(stringValue(binding["issued_at"]))
	if err != nil {
		return err
	}
	expiresAt, err := parseRFC3339String(stringValue(binding["expires_at"]))
	if err != nil {
		return err
	}
	if issuedAt.After(expiresAt) {
		return &Error{Message: "issued_at must not be later than expires_at"}
	}
	now := time.Now().UTC()
	if options.Now != "" {
		now, err = parseRFC3339String(options.Now)
		if err != nil {
			return err
		}
	}
	if now.Before(issuedAt) || now.After(expiresAt) {
		return &Error{Message: "binding is outside the accepted validity window"}
	}
	_, err = VerifyObjectProof(binding, agentDID, issuerDocument)
	return err
}

func validateDidWbaBindingFields(binding map[string]any) error {
	for _, field := range DidWbaBindingRequiredFields {
		value, _ := binding[field].(string)
		if value == "" {
			return &Error{Message: "missing proof field: " + field}
		}
	}
	if _, err := base64util.DecodeURL(stringValue(binding["leaf_signature_key_b64u"])); err != nil {
		return &Error{Message: "leaf_signature_key_b64u is not valid base64url data"}
	}
	if _, err := parseRFC3339String(stringValue(binding["issued_at"])); err != nil {
		return err
	}
	if _, err := parseRFC3339String(stringValue(binding["expires_at"])); err != nil {
		return err
	}
	return nil
}

File: rust/src/proof/did_wba_binding.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::authentication::{find_verification_method, is_assertion_method_authorized};
use crate::keys::base64url_decode;
use crate::proof::{parse_rfc3339, verify_object_proof, ProofError};
use crate::PrivateKeyMaterial;

use super::generate_object_proof;

pub const DID_WBA_BINDING_REQUIRED_FIELDS: [&str; 5] = [
    "agent_did",
    "verification_method",
    "leaf_signature_key_b64u",
    "issued_at",
    "expires_at",
];

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
pub struct DidWbaBindingVerificationOptions {
    pub now: Option<String>,
    pub expected_leaf_signature_key_b64u: Option<String>,
    pub expected_credential_identity: Option<String>,
}

pub fn generate_did_wba_binding(
    agent_did: &str,
    verification_method: &str,
    leaf_signature_key_b64u: &str,
    private_key: &PrivateKeyMaterial,
    issued_at: &str,
    expires_at: &str,
    proof_created: Option<String>,
) -> Result<Value, ProofError> {
    let binding = json!({
        "agent_did": agent_did,
        "verification_method": verification_method,
        "leaf_signature_key_b64u": leaf_signature_key_b64u,
        "issued_at": issued_at,
        "expires_at": expires_at,
    });
    validate_binding_fields(&binding)?;
    generate_object_proof(
        &binding,
        private_key,
        verification_method,
        agent_did,
        proof_created.or_else(|| Some(issued_at.to_string())),
    )
}

pub fn verify_did_wba_binding(
    binding: &Value,
    issuer_document: &Value,
    options: DidWbaBindingVerificationOptions,
) -> Result<(), ProofError> {
    validate_binding_fields(binding)?;
    let binding_object = binding.as_object().ok_or_else(|| {
        ProofError::InvalidProofField("did_wba_binding must be an object".to_string())
    })?;
    let agent_did = binding_object
        .get("agent_did")
        .and_then(Value::as_str)
        .ok_or_else(|| ProofError::MissingProofField("agent_did".to_string()))?;
    let verification_method = binding_object
        .get("verification_method")
        .and_then(Value::as_str)
        .ok_or_else(|| ProofError::MissingProofField("verification_method".to_string()))?;
    if verification_method.split('#').next() != Some(agent_did) {
        return Err(ProofError::IssuerDidMismatch);
    }
    if !is_assertion_method_authorized(issuer_document, verification_method) {
        return Err(ProofError::UnauthorizedVerificationMethod);
    }
    if find_verification_method(issuer_document, verification_method).is_none() {
        return Err(ProofError::InvalidProofField(
            "verification_method".to_string(),
        ));
    }
    if let Some(expected_leaf_key) = options.expected_leaf_signature_key_b64u.as_deref() {
        if binding_object
            .get("leaf_signature_key_b64u")
            .and_then(Value::as_str)
            != Some(expected_leaf_key)
        {
            return Err(ProofError::InvalidProofField(
                "leaf_signature_key_b64u".to_string(),
            ));
        }
    }
    if let Some(expected_identity) = options.expected_credential_identity.as_deref() {
        if expected_identity != agent_did {
            return Err(ProofError::InvalidProofField(
                "credential.identity".to_string(),
            ));
        }
    }

    let issued_at = parse_rfc3339(
        binding_object
            .get("issued_at")
            .and_then(Value::as_str)
            .ok_or_else(|| ProofError::MissingProofField("issued_at".to_string()))?,
    )?;
    let expires_at = parse_rfc3339(
        binding_object
            .get("expires_at")
            .and_then(Value::as_str)
            .ok_or_else(|| ProofError::MissingProofField("expires_at".to_string()))?,
    )?;
    if issued_at > expires_at {
        return Err(ProofError::InvalidProofField("expires_at".to_string()));
    }
    let now = match options.now.as_deref() {
        Some(value) => parse_rfc3339(value)?,
        None => Utc::now(),
    };
    ensure_time_window(now, issued_at, expires_at)?;

    verify_object_proof(binding, agent_did, issuer_document)
}

fn validate_binding_fields(binding: &Value) -> Result<(), ProofError> {
    let object = binding.as_object().ok_or_else(|| {
        ProofError::InvalidProofField("did_wba_binding must be an object".to_string())
    })?;
    for field in DID_WBA_BINDING_REQUIRED_FIELDS {
        let value = object
            .get(field)
            .and_then(Value::as_str)
            .unwrap_or_default();
        if value.is_empty() {
            return Err(ProofError::MissingProofField(field.to_string()));
        }
    }
    let leaf_key = object
        .get("leaf_signature_key_b64u")
        .and_then(Value::as_str)
        .ok_or_else(|| ProofError::MissingProofField("leaf_signature_key_b64u".to_string()))?;
    base64url_decode(leaf_key)
        .map_err(|_| ProofError::InvalidProofField("leaf_signature_key_b64u".to_string()))?;
    parse_rfc3339(
        object
            .get("issued_at")
            .and_then(Value::as_str)
            .ok_or_else(|| ProofError::MissingProofField("issued_at".to_string()))?,
    )?;
    parse_rfc3339(
        object
            .get("expires_at")
            .and_then(Value::as_str)
            .ok_or_else(|| ProofError::MissingProofField("expires_at".to_string()))?,
    )?;
    Ok(())
}

fn ensure_time_window(
    now: DateTime<Utc>,
    issued_at: DateTime<Utc>,
    expires_at: DateTime<Utc>,
) -> Result<(), ProofError> {
    if now < issued_at || now > expires_at {
        return Err(ProofError::VerificationFailed);
    }
    Ok(())
}
